### ISSUE TITLE
Error linking in release mode with libc with specific code

### ISSUE THREAD

**Issue by jeffkdev** [ziglang/zig#9076]
With the release of 0.8.0 most of my test programs give the same error when linking in release mode, likely due to them sharing the same problematic code path. Otherwise in debug mode (and other release mode tests with libc) have been working fine.

Release fast results:
```zig build-exe link_error_example.zig some_source_file.c -lc -OReleaseFast``` 
```
lld-link: warning: ignoring debug info with an invalid version (0) in zig-cache\o\cce9e6bd14907d41ac812c3ec0993559\link_error_example.obj
error: LLDReportedFailure
```

Release safe results:
```zig build-exe link_error_example.zig some_source_file.c -lc -OReleaseSafe```
```
lld-link: warning: ignoring debug info with an invalid version (0) in zig-cache\o\779e573390a38de497af37409173322a\link_error_example.obj
lld-link: error: undefined symbol: _tls_index
>>> referenced by link_error_example.exe.lto.obj:(std.debug.panicExtra)
>>> referenced by link_error_example.exe.lto.obj:(std.debug.panicExtra)
>>> referenced by link_error_example.exe.lto.obj:(std.debug.panicExtra)
error: LLDReportedFailure
```

I isolated the problem down to:
- Must be in release mode (fast, small, or safe)
- Must link libC
- Must link any arbitrary c file (can be empty)
- Use this 60 line zig example file below

This is the smallest example I was able to reproduce the bug with (entirely meaningless code at this point).
```zig
const std = @import("std");

pub const Vec1 = struct {
    x: f32,
    pub const zero = Vec1{ .x = 0 };
};

pub const Vec3 = struct {
    const Self = @This();
    pub const zero = Self.new(0, 0, 0);

    x: f32,
    y: f32,
    z: f32,

    pub fn new(x: f32, y: f32, z: f32) Self {
        return Self{ .x = x, .y = y, .z = z };
    }
    pub fn add(a: Self, b: Self) Self {
        return new(a.x + b.x, a.y + b.y, a.z + b.z);
    }
    pub fn scl(a: Self, b: f32) Self {
        return new(a.x * b, a.y * b, a.z * b);
    }
};

pub const Data = struct {
    orbital: Vec3 = Vec3.zero,
    center: Vec3 = Vec3.zero,
};

pub const View = struct {
    offset_y: f32 = 0.0,
    eyePos: Vec3 = Vec3.zero,
    proj: Vec1 = Vec1.zero,
};

var camera_data: Data = Data{};
var camera_view: View = View{};

pub fn main() void {
    camera_data.orbital.y = std.math.max(std.math.min(camera_data.orbital.y, 0), 1);
    camera_view.proj = Vec1.zero;
    const center = camera_data.center.add(Vec3.new(0.0, camera_view.offset_y, 0.0));

    const latitude = camera_data.orbital.y;
    const longitude = camera_data.orbital.x;
    const some_value = Vec3{ .x = @cos(latitude) * @sin(longitude), .y = @sin(latitude), .z = 0 };
    camera_view.eyePos = some_value.scl(1).add(center);
}
```
And then it must also include an arbitrary c file:
some_source_file.c
```
// Doesn't matter what is in here... previously this file included actual files
```

OS: Windows 10
Zig version: 0.9.0-dev.70+a3f7a48d9, built from source

Here is a zip of the zig file, an empty c file and the build command (same as above):
[link_error_example.zip](https://github.com/ziglang/zig/files/6635445/link_error_example.zip)

**Comment by LemonBoy** [comment]
You must really hate the search function because this is a duplicate of #8531, the solution is building windows libc (or only the file exporting the TLS symbols required by the runtime) without LTO.

**Comment by jeffkdev** [comment]
It looks like in an attempt to create a minimal case I might have oversimplified the issue. It does fail using the command line arguments I provided, but not in the strange way I was seeing. Instead, when using this build.zig file even changing an arbitrary parameter from a 1 to a 0 changes the results. I got to the point where removing any line would allow it to build. For example
```camera_data.orbital.y = std.math.max(std.math.min(camera_data.orbital.y, 0), 1); // error```
to
```camera_data.orbital.y = std.math.max(std.math.min(camera_data.orbital.y, 0), 0); //success```
I also just noticed that it did briefly show a warning even in the case of a success:
```
warning(link): unexpected LDD stderr:
ldd-link: warning: ignoring debug info with an invalid version (0) in [truncated]\src\zig-cache-o\[long_string]\program.obj
```

build.zig
```zig
const std = @import("std");
const builtin = @import("builtin");

pub fn build(b: *std.build.Builder) anyerror!void {
    b.release_mode = .ReleaseFast;
    b.is_release = true;
    const mode = b.standardReleaseOptions();
    var exe = b.addExecutable("program", "link_error_example.zig");
    exe.setBuildMode(mode);
    exe.addCSourceFile("some_source_file.c", &.{});
    exe.linkLibC();
    const run_cmd = exe.run();
    const run_step = b.step("run", "Run the app");
    run_step.dependOn(&run_cmd.step);
    b.default_step.dependOn(&exe.step);
    b.installArtifact(exe);
}
```
It produces the command string:
`zig.exe build-exe [truncated]\src\link_error_example.zig  -- [truncated]\src\some_source_file.c -lc -OReleaseFast --cache-dir [truncated]\src\zig-cache --global-cache-dir [truncated]\AppData\Local\zig --name program --enable-cache`
Using that method I am able to reproduce the "only works with very specific file" condition. However, the command in the first post will always fail even after making trivial changes.

In the general case I am able to "Mix Zig and C sources on Windows in ReleaseSafe mode" without causing linker errors, for example these all run fine:  https://github.com/jeffkdev/sokol-zig-examples which was why when I searched "LLDReportedFailure" and saw the 20 open issues, the one you linked didn't stand out to me as being the same. 

But thanks for pointing me in the right direction, using `-fno-lto` solved my immediate problem. Does it make sense that trivial changes to the zig code would make the LTO issues go away, but only when using the named caches provided by the build.zig script?
